packages/firestore/test/unit/remote/bloom_filter.test.ts

```diff
@@ -0,0 +1,393 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { expect } from 'chai';
+
+import { BloomFilter } from '../../../src/remote/bloom_filter';
+import { ByteString } from '../../../src/util/byte_string';
+
+import {
+  testDataCount1Rate0001,
+  testDataCount1Rate01,
+  testDataCount1Rate1,
+  testDataCount50000Rate0001,
+  testDataCount50000Rate01,
+  testDataCount50000Rate1,
+  testDataCount5000Rate0001,
+  testDataCount5000Rate01,
+  testDataCount5000Rate1,
+  testDataCount500Rate0001,
+  testDataCount500Rate01,
+  testDataCount500Rate1,
+  testResultCount1Rate0001,
+  testResultCount1Rate01,
+  testResultCount1Rate1,
+  testResultCount50000Rate0001,
+  testResultCount50000Rate01,
+  testResultCount50000Rate1,
+  testResultCount5000Rate0001,
+  testResultCount5000Rate01,
+  testResultCount5000Rate1,
+  testResultCount500Rate0001,
+  testResultCount500Rate01,
+  testResultCount500Rate1
+} from './bloom_filter_golden_test_data';
+
+describe('BloomFilter', () => {
+  it('can initiate an empty bloom filter', () => {
+    const bloomFilter = new BloomFilter(new Uint8Array(0), 0, 0);
+    expect(bloomFilter.size).to.equal(0);
+  });
+
+  it('empty bloom filter can have positive hash count', () => {
+    const bloomFilter = new BloomFilter(new Uint8Array(0), 0, 1);
+    expect(bloomFilter.size).to.equal(0);
+  });
+
+  it('should throw error if empty bloom filter inputs are invalid', () => {
+    try {
+      new BloomFilter(new Uint8Array(0), 1, 0);
+      expect.fail();
+    } catch (error) {
+      expect(
+        (error as Error)?.message.includes(
+          'INTERNAL ASSERTION FAILED: A valid empty bloom filter should have 0 padding.'
+        )
+      ).to.be.true;
+    }
+  });
+
+  it('can initiate a non empty bloom filter', () => {
+    const bloomFilter = new BloomFilter(
+      new Uint8Array([151, 153, 236, 116, 7]),
+      3,
+      13
+    );
+    expect(bloomFilter.size).to.equal(37);
+  });
+
+  it('should throw error if padding is negative', () => {
+    try {
+      new BloomFilter(new Uint8Array(1), -1, 1);
+      expect.fail();
+    } catch (error) {
+      expect(
+        (error as Error)?.message.includes(
+          'INTERNAL ASSERTION FAILED: Padding is negative.'
+        )
+      ).to.be.true;
+    }
+  });
+
+  it('should throw error if bitmap size is negative', () => {
+    try {
+      new BloomFilter(new Uint8Array(1), 9, 1);
+      expect.fail();
+    } catch (error) {
+      expect(
+        (error as Error)?.message.includes(
+          'INTERNAL ASSERTION FAILED: Bitmap size is negative.'
+        )
+      ).to.be.true;
+    }
+  });
+
+  it('should throw error if hash count is negative', () => {
```

Comment:
By the way, these test names are fantastic: they describe what behavior the test is verifying, which is very readable.

Reply:
😊